File: practise/five.rb
# program to find the Greatest Common Divisor (GCD) of two integers

def gcd(a, b)
	gcd = 1
	i = [a.abs, b.abs].min
	while i >= 1
		if a % i == 0 && b % i == 0
			gcd = i
			break
		end
		i -= 1
	end
	gcd
end

def better_gcd(a, b)
	a = a.abs
	b = b.abs
	return a if b == 0
	while a != b
		if a > b
			a -= b
		else
			b -= a
		end
	end
	a
end

def optimal_gcd(a, b)
	a = a.abs
	b = b.abs
	return a if b == 0
	optimal_gcd(b, a % b)
end

x = 10
y = 20
# puts "GCD of number #{x} and #{y} is #{gcd(x, y)}"
# puts "Better GCD of number #{x} and #{y} is #{better_gcd(x, y)}"
puts "Optimal GCD of number #{x} and #{y} is #{optimal_gcd(x, y)}"

File: practise/five.cpp
// program to finding the Greatest Common Divisor (GCD), or Greatest Common Factor (GCF), of two integers.
#include<bits/stdc++.h>
using namespace std;

// BRUTE FORCE SOLUTION
int gcd(int a, int b){
	int gcd = 1;
	for(int i=min(a,b); i>=1; i--){
		if(a%i == 0 && b%i == 0){
			gcd = i;
			break;
		}
	}
	return gcd;
}


// BETTER SOLUTIONS
int better_gcd(int a, int b){
	while(a != b){
		if(a > b) a -= b;
		else b -= a;
	}
	return a;
}

// OPTIMAL SOLUTIONS -->  Euclid’s Algorithm

int optimal_gcd(int a, int b){
	if (b == 0) return a;
	return optimal_gcd(b, a % b);

}

int main()
{
	int x = 10;
	int y = 20;
    // cout << "GCD of number " << x << " and " << y << " is " <<  gcd(x,y) << endl;
    // cout << "Better GCD of number " << x << " and " << y << " is " <<  better_gcd(x,y) << endl;
    cout << "Optimal GCD of number " << x << " and " << y << " is " <<  optimal_gcd(x,y) << endl;
	return 0;
}
